{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate reference dataframes\n",
    "\n",
    "Generate dataframe object containing lncRNA gtf annotations with genomic cooridnates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "infile=\"reference-annotation/lncipedia_5_2_hg38.gtf\"\n",
    "gtf_init=[]\n",
    "with open(infile) as gtf:\n",
    "    for _ in range(2):         #Skips header lines\n",
    "        next(gtf)\n",
    "    for line in gtf:\n",
    "        gtf_init.append(line)\n",
    "ref_gtf=pd.DataFrame([entry.strip().split('\\t') for entry in gtf_init], \n",
    "                     columns=('contig', 'source','feature','start','end','score','strand','frame','attribute'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate dataframe object containing genome gtf annotation specifically for 'gene' feature types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_ref=\"reference-annotation/hg38.gtf\"\n",
    "ref_init=[]\n",
    "with open(coding_ref) as ref:\n",
    "    for _ in range(5):          #Skips header lines\n",
    "        next(ref)\n",
    "    for line in ref:\n",
    "        temp = line.strip().split('\\t')\n",
    "        if temp[2] == \"gene\":\n",
    "            ref_init.append(temp)\n",
    "        else:\n",
    "            pass\n",
    "ref_coding=pd.DataFrame(ref_init,columns=('contig', 'source','feature','start','end','score','strand','frame','attribute'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate dataframe object containing CNV information from DepMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_input = \"reference-annotation/CCLE_ABSOLUTE_combined_20181227.txt\"\n",
    "cnv_ref = pd.read_csv(cnv_input, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse input tsv data containing genes and cell lines of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "input_genes = \"./example-gene-input.tsv\"\n",
    "input_depmap_ID = \"./example-depmap_ID-input.tsv\"\n",
    "\n",
    "gene_list = pd.read_csv(input_genes,sep = '\\t')\n",
    "depmap_ID_list = pd.read_csv(input_depmap_ID, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test for overlaps with alternative genes and regions containing known CNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: SLX1B-SULT1A4 could not be found within the LNCipedia reference data\n",
      "1 input gene(s) could not be found within the LNCipedia reference data\n",
      "No overlapping CNV regions were identified against input genes for all input depmap IDs\n"
     ]
    }
   ],
   "source": [
    "output_prefix= \"test\"\n",
    "missing_genes = 0\n",
    "total_overlap_genes = pd.DataFrame(columns=['input_gene','contig','source','feature','start','end','score','strand','frame','attribute'])\n",
    "total_overlap_cnv = pd.DataFrame(columns=['input_gene','sample',\n",
    " 'Chromosome',\n",
    " 'Start',\n",
    " 'End',\n",
    " 'Num_Probes',\n",
    " 'Length',\n",
    " 'Modal_HSCN_1',\n",
    " 'Modal_HSCN_2',\n",
    " 'Modal_Total_CN',\n",
    " 'Subclonal_HSCN_a1',\n",
    " 'Subclonal_HSCN_a2',\n",
    " 'Cancer_cell_frac_a1',\n",
    " 'Ccf_ci95_low_a1',\n",
    " 'Ccf_ci95_high_a1',\n",
    " 'Cancer_cell_frac_a2',\n",
    " 'Ccf_ci95_low_a2',\n",
    " 'Ccf_ci95_high_a2',\n",
    " 'LOH',\n",
    " 'Homozygous_deletion',\n",
    " 'depMapID'])\n",
    "for x in range(len(gene_list)):\n",
    "    gene = gene_list.iloc[x]['gene']\n",
    "    try:\n",
    "        # Generates start and end positions for the lncRNA gene\n",
    "        import re\n",
    "        df=ref_gtf[ref_gtf['attribute'].str.contains(gene)]\n",
    "        gene_chrom = re.findall(r'\\d+', df.iloc[0]['contig'])\n",
    "        gene_start = int(df.iloc[0]['start'])\n",
    "        gene_end = int(df.iloc[-1]['end'])\n",
    "        # Checks whether lncRNA overlaps with any other genes\n",
    "        overlaps_init = []\n",
    "        for i in range(len(ref_coding)):\n",
    "            if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "                if gene_start <= int(ref_coding['start'][i]) <= gene_end:\n",
    "                    overlaps_init.append(ref_coding.iloc[i])\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "            if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "                if gene_start <= int(ref_coding['end'][i]) <= gene_end:\n",
    "                    overlaps_init.append(ref_coding.iloc[i])\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "        overlaps = pd.DataFrame(overlaps_init)\n",
    "        overlaps = pd.DataFrame.drop_duplicates(overlaps)\n",
    "        overlaps.insert(0,\"input_gene\", gene)\n",
    "        total_overlap_genes=pd.concat([total_overlap_genes,overlaps])\n",
    "        \n",
    "        aggregate_overlaps=pd.DataFrame(columns=list(cnv_ref.columns))\n",
    "        for y in range(len(depmap_ID_list)):\n",
    "            depmap_ID = depmap_ID_list.iloc[y]['depmap_ID']\n",
    "            # Checks whether lncRNA overlaps with known CNV region\n",
    "            cnv_init = []\n",
    "            filter_cnv = cnv_ref[cnv_ref['depMapID']==depmap_ID]\n",
    "            for i in range(len(filter_cnv)):\n",
    "                if filter_cnv.iloc[i]['Chromosome'] == int(gene_chrom[0]):\n",
    "                    if gene_start <= int(filter_cnv.iloc[i]['Start']) <= gene_end:\n",
    "                        cnv_init.append(filter_cnv.iloc[i])\n",
    "                    else:\n",
    "                        pass\n",
    "                    if gene_start <= int(filter_cnv.iloc[i]['End']) <= gene_end:\n",
    "                        cnv_init.append(filter_cnv.iloc[i])\n",
    "                    else:\n",
    "                        pass\n",
    "                else:\n",
    "                    pass\n",
    "            cnv_overlap = pd.DataFrame(cnv_init)\n",
    "            cnv_overlap = pd.DataFrame.drop_duplicates(cnv_overlap)\n",
    "            aggregate_overlaps=pd.concat([aggregate_overlaps,cnv_overlap])\n",
    "        total_overlap_cnv=pd.concat([total_overlap_cnv,aggregate_overlaps])\n",
    "    except:\n",
    "        print(\"Warning: \" + gene + \" could not be found within the LNCipedia reference data\")\n",
    "        missing_genes += 1\n",
    "if total_overlap_genes.empty:\n",
    "    print(\"No overlapping genes were identified against input genes\")\n",
    "else:\n",
    "    pd.DataFrame.to_csv(total_overlap_genes, sep='\\t', path_or_buf= output_prefix+\"_overlapping_genes.tsv\", index=False)\n",
    "print(str(missing_genes) + \" input gene(s) could not be found within the LNCipedia reference data\")\n",
    "if total_overlap_cnv.empty:\n",
    "    print(\"No overlapping CNV regions were identified against input genes for all input depmap IDs\")\n",
    "else:\n",
    "    pd.DataFrame.to_csv(cnv_overlap, sep='\\t', path_or_buf= output_prefix+\"_overlapping_cnv.tsv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Full run code\n",
    "Input variables are denoted, with comments providing additional details. The code generates 2 potential .tsv files as outputs for any given gene:\n",
    "1.  _\"overlapping_genes.tsv\"_ returns the gtf entries of genes that overlap with the genomic coordinates of the lncRNA gene of interest\n",
    "\n",
    "2.  _\"overlapping_cnv.tsv\"_ return the entries of CNV regions that overlap with the genomic coordinates of the lncRNA gene of interest, keeping with the format of the original depmap CNV reference annotation\n",
    "\n",
    "__These outputs are not generated if no relevant overlaps are detected__ \n",
    "\n",
    "In the case that the lncRNA gene of interest cannot be found in the lncRNA reference data, a warning message will appear denoting the specific lncRNA gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: SLX1B-SULT1A4 could not be found within the LNCipedia reference data\n"
     ]
    }
   ],
   "source": [
    "### Adjust the following input parameters as required\n",
    "############################################################################\n",
    "input_lncrna_ref = \"reference-annotation/lncipedia_5_2_hg38.gtf\"           # A reference gtf file specifying the genomic coordinates for your lncRNAs of interest; gene names must be present in the attributes column\n",
    "lncrna_header = 2                                                          # The number of header/comment rows prior to gtf entries in \"input_lncrna_ref\"\n",
    "input_coding_ref = \"reference-annotation/hg38.gtf\"                         # A reference gtf file containing the genomic coordinates of protein coding genes e.g. hg38 reference gtf\n",
    "coding_header = 5                                                          # The number of header/comment rows prior to gtf entries in \"input_coding_ref\"\n",
    "input_cnv_ref = \"reference-annotation/CCLE_ABSOLUTE_combined_20181227.txt\" # Reference data downloaded from depmap containing CNV annotation \"https://depmap.org/portal/download/all/?release=CCLE+2019&file=CCLE_ABSOLUTE_combined_20181227.xlsx\"                                                                      \n",
    "input_genes = \"./example-gene-input.tsv\"                                   # A tsv file with one column named \"gene\" listing all lncRNA genes of interest\n",
    "input_depmap_ID = \"./example-depmap_ID-input.tsv\"                          # A tsv file with one column named \"depmap_ID\" listing the depmap IDs for all cell lines of interest\n",
    "output_prefix= \"lncRNA\"                                                    # Prefix to be specified for any output files generated\n",
    "############################################################################\n",
    "\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "gtf_init=[]\n",
    "with open(input_lncrna_ref) as gtf:\n",
    "    for _ in range(lncrna_header):\n",
    "        next(gtf)\n",
    "    for line in gtf:\n",
    "        gtf_init.append(line)\n",
    "ref_gtf=pd.DataFrame([entry.strip().split('\\t') for entry in gtf_init], \n",
    "                     columns=('contig', 'source','feature','start','end','score','strand','frame','attribute'))\n",
    "\n",
    "\n",
    "ref_init=[]\n",
    "with open(input_coding_ref) as ref:\n",
    "    for _ in range(coding_header):\n",
    "        next(ref)\n",
    "    for line in ref:\n",
    "        temp = line.strip().split('\\t')\n",
    "        if temp[2] == \"gene\":\n",
    "            ref_init.append(temp)\n",
    "        else:\n",
    "            pass\n",
    "ref_coding=pd.DataFrame(ref_init,columns=('contig', 'source','feature','start','end','score','strand','frame','attribute'))\n",
    "\n",
    "cnv_ref = pd.read_csv(input_cnv_ref, sep = '\\t')\n",
    "\n",
    "gene_list = pd.read_csv(input_genes,sep = '\\t')\n",
    "depmap_ID_list = pd.read_csv(input_depmap_ID, sep = '\\t')\n",
    "\n",
    "missing_genes = 0\n",
    "total_overlap_genes = pd.DataFrame(columns=['input_gene','contig','source','feature','start','end','score','strand','frame','attribute'])\n",
    "total_overlap_cnv = pd.DataFrame(columns=['input_gene','sample',\n",
    " 'Chromosome',\n",
    " 'Start',\n",
    " 'End',\n",
    " 'Num_Probes',\n",
    " 'Length',\n",
    " 'Modal_HSCN_1',\n",
    " 'Modal_HSCN_2',\n",
    " 'Modal_Total_CN',\n",
    " 'Subclonal_HSCN_a1',\n",
    " 'Subclonal_HSCN_a2',\n",
    " 'Cancer_cell_frac_a1',\n",
    " 'Ccf_ci95_low_a1',\n",
    " 'Ccf_ci95_high_a1',\n",
    " 'Cancer_cell_frac_a2',\n",
    " 'Ccf_ci95_low_a2',\n",
    " 'Ccf_ci95_high_a2',\n",
    " 'LOH',\n",
    " 'Homozygous_deletion',\n",
    " 'depMapID'])\n",
    "for x in range(len(gene_list)):\n",
    "    gene = gene_list.iloc[x]['gene']\n",
    "    try:\n",
    "        # Generates start and end positions for the lncRNA gene\n",
    "        import re\n",
    "        df=ref_gtf[ref_gtf['attribute'].str.contains(gene)]\n",
    "        gene_chrom = re.findall(r'\\d+', df.iloc[0]['contig'])\n",
    "        gene_start = int(df.iloc[0]['start'])\n",
    "        gene_end = int(df.iloc[-1]['end'])\n",
    "        # Checks whether lncRNA overlaps with any other genes\n",
    "        overlaps_init = []\n",
    "        for i in range(len(ref_coding)):\n",
    "            if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "                if gene_start <= int(ref_coding['start'][i]) <= gene_end:\n",
    "                    overlaps_init.append(ref_coding.iloc[i])\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "            if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "                if gene_start <= int(ref_coding['end'][i]) <= gene_end:\n",
    "                    overlaps_init.append(ref_coding.iloc[i])\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "        overlaps = pd.DataFrame(overlaps_init)\n",
    "        overlaps = pd.DataFrame.drop_duplicates(overlaps)\n",
    "        overlaps.insert(0,\"input_gene\", gene)\n",
    "        total_overlap_genes=pd.concat([total_overlap_genes,overlaps])\n",
    "        \n",
    "        aggregate_overlaps=pd.DataFrame(columns=list(cnv_ref.columns))\n",
    "        for y in range(len(depmap_ID_list)):\n",
    "            depmap_ID = depmap_ID_list.iloc[y]['depmap_ID']\n",
    "            # Checks whether lncRNA overlaps with known CNV region\n",
    "            cnv_init = []\n",
    "            filter_cnv = cnv_ref[cnv_ref['depMapID']==depmap_ID]\n",
    "            for i in range(len(filter_cnv)):\n",
    "                if filter_cnv.iloc[i]['Chromosome'] == int(gene_chrom[0]):\n",
    "                    if gene_start <= int(filter_cnv.iloc[i]['Start']) <= gene_end:\n",
    "                        cnv_init.append(filter_cnv.iloc[i])\n",
    "                    else:\n",
    "                        pass\n",
    "                    if gene_start <= int(filter_cnv.iloc[i]['End']) <= gene_end:\n",
    "                        cnv_init.append(filter_cnv.iloc[i])\n",
    "                    else:\n",
    "                        pass\n",
    "                else:\n",
    "                    pass\n",
    "            cnv_overlap = pd.DataFrame(cnv_init)\n",
    "            cnv_overlap = pd.DataFrame.drop_duplicates(cnv_overlap)\n",
    "            aggregate_overlaps=pd.concat([aggregate_overlaps,cnv_overlap])\n",
    "        total_overlap_cnv=pd.concat([total_overlap_cnv,aggregate_overlaps])\n",
    "    except:\n",
    "        print(\"Warning: \" + gene + \" could not be found within the LNCipedia reference data\")\n",
    "        missing_genes += 1\n",
    "if total_overlap_genes.empty:\n",
    "    print(\"No overlapping genes were identified against input genes\")\n",
    "else:\n",
    "    pd.DataFrame.to_csv(total_overlap_genes, sep='\\t', path_or_buf= output_prefix+\"_overlapping_genes.tsv\", index=False)\n",
    "print(str(missing_genes) + \" input gene(s) could not be found within the LNCipedia reference data\")\n",
    "if total_overlap_cnv.empty:\n",
    "    print(\"No overlapping CNV regions were identified against input genes for all input depmap IDs\")\n",
    "else:\n",
    "    pd.DataFrame.to_csv(cnv_overlap, sep='\\t', path_or_buf= output_prefix+\"_overlapping_cnv.tsv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Archived code\n",
    "Original code which checks for overlaps with other genes and CNV regions based on a manually entered gene and cell depmap_ID. This can still be useful if just wishing to quickly check a single gene and cell line combination. Code used in the full run reads from two .tsv input files: a list of genes and a list of depmap_IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "depmap_ID = \"ACH-000045\"    # Cell-line ID as found on depmap which is used as the source for CNV reference info\n",
    "gene =\"RP1-283E3.8\"         # Replace gene variable with gene of interest\n",
    "\n",
    "# Generates start and end positions for the lncRNA gene\n",
    "import re\n",
    "df=ref_gtf[ref_gtf['attribute'].str.contains(gene)]\n",
    "gene_chrom = re.findall(r'\\d+', df.iloc[0]['contig'])\n",
    "gene_start = int(df.iloc[0]['start'])\n",
    "gene_end = int(df.iloc[-1]['end'])\n",
    "\n",
    "# Checks whether lncRNA overlaps with any other genes\n",
    "overlaps_init = []\n",
    "for i in range(len(ref_coding)):\n",
    "    if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "        if gene_start <= int(ref_coding['start'][i]) <= gene_end:\n",
    "            overlaps_init.append(ref_coding.iloc[i])\n",
    "        else:\n",
    "            pass\n",
    "    else:\n",
    "        pass\n",
    "    if ref_coding['contig'][i] == gene_chrom[0]:\n",
    "        if gene_start <= int(ref_coding['end'][i]) <= gene_end:\n",
    "            overlaps_init.append(ref_coding.iloc[i])\n",
    "        else:\n",
    "            pass\n",
    "    else:\n",
    "        pass\n",
    "overlaps = pd.DataFrame(overlaps_init)\n",
    "overlaps = pd.DataFrame.drop_duplicates(overlaps)\n",
    "\n",
    "# Checks whether lncRNA overlaps with known CNV region\n",
    "cnv_init = []\n",
    "filter_cnv = cnv_ref[cnv_ref['depMapID']==depmap_ID]\n",
    "for i in range(len(filter_cnv)):\n",
    "    if filter_cnv.iloc[i]['Chromosome'] == int(gene_chrom[0]):\n",
    "        if gene_start <= int(filter_cnv.iloc[i]['Start']) <= gene_end:\n",
    "            cnv_init.append(filter_cnv.iloc[i])\n",
    "        else:\n",
    "            pass\n",
    "        if gene_start <= int(filter_cnv.iloc[i]['End']) <= gene_end:\n",
    "            cnv_init.append(filter_cnv.iloc[i])\n",
    "        else:\n",
    "            pass\n",
    "    else:\n",
    "        pass\n",
    "cnv_overlap = pd.DataFrame(cnv_init)\n",
    "cnv_overlap = pd.DataFrame.drop_duplicates(cnv_overlap)\n",
    "\n",
    "print(overlaps.head)\n",
    "print(cnv_overlap.head)\n",
    "\n",
    "# pd.DataFrame.to_csv(overlaps, sep='\\t', path_or_buf=gene+\"_overlapping_genes.tsv\", index=False)\n",
    "# pd.DataFrame.to_csv(cnv_overlap, sep='\\t', path_or_buf=gene+\"_\"+ depmap_ID +\"_overlapping_cnv.tsv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Positive control code demonstrating whether overlap with regions of CNV could be identified for all depmap_IDs for a given gene of interest. The ```gene_start``` and ```gene_end``` values are artificially inputted to cause overlaps "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\she\\AppData\\Local\\Temp\\ipykernel_14088\\3944915674.py:27: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  aggregate_overlaps=pd.concat([aggregate_overlaps,cnv_overlap])\n"
     ]
    }
   ],
   "source": [
    "gene =\"RP1-283E3.8\" \n",
    "df=ref_gtf[ref_gtf['attribute'].str.contains(gene)]\n",
    "gene_chrom = re.findall(r'\\d+', df.iloc[0]['contig'])\n",
    "gene_start=247741422\n",
    "gene_end=248562279\n",
    "columns = list(cnv_ref.columns)\n",
    "aggregate_overlaps=pd.DataFrame(columns=columns)\n",
    "for y in range(len(depmap_ID_list)):\n",
    "    depmap_ID = depmap_ID_list.iloc[y]['depmap_ID']\n",
    "    # Checks whether lncRNA overlaps with known CNV region\n",
    "    cnv_init = []\n",
    "    filter_cnv = cnv_ref[cnv_ref['depMapID']==depmap_ID]\n",
    "    for i in range(len(filter_cnv)):\n",
    "        if filter_cnv.iloc[i]['Chromosome'] == int(gene_chrom[0]):\n",
    "            if gene_start <= int(filter_cnv.iloc[i]['Start']) <= gene_end:\n",
    "                cnv_init.append(filter_cnv.iloc[i])\n",
    "            else:\n",
    "                pass\n",
    "            if gene_start <= int(filter_cnv.iloc[i]['End']) <= gene_end:\n",
    "                cnv_init.append(filter_cnv.iloc[i])\n",
    "            else:\n",
    "                pass\n",
    "        else:\n",
    "            pass\n",
    "    cnv_overlap = pd.DataFrame(cnv_init)\n",
    "    cnv_overlap = pd.DataFrame.drop_duplicates(cnv_overlap)\n",
    "    aggregate_overlaps=pd.concat([aggregate_overlaps,cnv_overlap])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
